src/transform/src/monotonic.rs

```diff
@@ -10,214 +10,45 @@
 //! Analysis to identify monotonic collections, especially TopK inputs.
 use std::collections::BTreeSet;
 
-use itertools::zip_eq;
-use mz_expr::visit::VisitChildren;
-use mz_expr::{Id, LetRecLimit, LocalId, MirRelationExpr, RECURSION_LIMIT};
-use mz_ore::cast::CastFrom;
-use mz_ore::soft_panic_or_log;
-use mz_ore::stack::{CheckedRecursion, RecursionGuard};
+use crate::analysis::monotonic::Monotonic;
+use crate::analysis::DerivedBuilder;
+use crate::TransformCtx;
+use mz_expr::MirRelationExpr;
 use mz_repr::GlobalId;
 
 /// A struct that holds a recursive function that determines if a
 /// relation is monotonic, and applies any optimizations along the way.
-#[derive(Debug)]
-pub struct MonotonicFlag {
-    recursion_guard: RecursionGuard,
-}
-
-impl Default for MonotonicFlag {
-    fn default() -> MonotonicFlag {
-        MonotonicFlag {
-            recursion_guard: RecursionGuard::with_limit(RECURSION_LIMIT),
-        }
-    }
-}
-
-impl CheckedRecursion for MonotonicFlag {
-    fn recursion_guard(&self) -> &RecursionGuard {
-        &self.recursion_guard
-    }
-}
+#[derive(Debug, Default)]
+pub struct MonotonicFlag;
 
 impl MonotonicFlag {
-    /// Determines if a relation is monotonic, and applies any optimizations along the way.
-    /// mon_ids should be the ids of monotonic sources and indexes involved in expr.
-    pub fn apply(
+    ///
```

Comment:
Missing comment :)

Reply:
(Note that the PR was still in an explicitly marked draft state.)